docs/support-for-generic-function-types-in-user-defined-type-functions.md

```diff
@@ -0,0 +1,120 @@
+We also have a different representation of a table with a metatable between type functions and typechecking.
+
+Generics are created by specifying a name and a boolean marking it as a type or a pack.
+Generics that are coming from type inference might not have a display name, but still hold an auto-generated name internally.
+
+Names are the way generic identity is specified.
+Generic types are immutable and two generics with the same name are equal, even if they are defined in different function type scopes.
+
+For example, consider a type `<T, U>(T, U, <T>(T) -> U) -> ()`
+
+From the typechecking perspective, inner and outer function 'T' types are different, but they are still represented by the same runtime generic named 'T'.
+
+The ambiguity is resolved based on the scoping of the generic in question.
+When the runtime type is returned back to the typechecking engine, only those generics that are in the current set of function type generic lists are allowed to be mentioned.
+When the inner generic function list places the name 'T' again, it represents a different typechecking type and hides the outer 'T' until the scope of the inner type ends.
+
+So we can say generics are defined by the lexical scope, which matches how they are written out in source code.
+
+Just like in the syntax of the language, reusing the same generic name, even between types and packs in a single list results in a duplicate generic definition error.
+
+### Examples
+
+Creating a function type with generic types and packs:
+
+```lua
+type function pass()
+    local T = types.generic("T")
+    local Us, Vs = types.generic("U", true), types.generic("V", true)
+
+    local f = types.newfunction()
+    f:setparameters({T}, Us);
+    f:setreturns({T}, Vs);
+    f:setgenerics({T, Us, Vs});
+    return f
+end
+
+type X = pass<> -- <T, U..., V...>(T, U...) -> (T, V...)
+```
+
+Creating a function type with a nested function type that reuses a generic name:
+
+```lua
+type function pass()
+    local T, U = types.generic("T"), types.generic("U")
+
+    -- <T>(T) -> ()
+    local func = types.newfunction({ head = {T} }, {}, {T});
+
+    -- { x: <T>(T) -> (), y: U }
+    local tbl = types.newtable({ [types.singleton("x")] = func, [types.singleton("y")] = U })
+
+    -- <T, U>(T, { x: <T>(T) -> (), y: U }) -> ()
+    return types.newfunction({ head = {T, tbl} }, {}, {T, U})
+end
+```
+
+## Updates to the library and type userdata
+
+### `types` Library
+
+| New/Update | Library Functions | Return Type | Description |
+| ------------- | ------------- | ------------- | ------------- |
+| New |  `generic(name: string, ispack: boolean?)` | `type` | returns an immutable instance of a generic type or a pack; name cannot be empty |
+| Update | `newfunction(parameters: { head: {type}?, tail: type? }, returns: { head: {type}?, tail: type? }, generics: {type}?)` | `type` | New `generic` argument is added to specify generic types and packs in a single list |
```

Comment:
This is sort of unfortunate parameter ordering, but at the same time, I think it's probably the one we'd pick based on usefulness even if we had generics all along.

Reply:
Doesn't exactly follow the order we have when writing source code, but making a 3-arg overload with a different order seemed like a path to confusion.